File: Behavioral/Interpreter/cpp/test_interpreter.cpp

#include "interpreter.hpp"
#include <gtest/gtest.h>

using namespace interpreter_pattern;

/**
 * @brief Test evaluating constants and variables.
 */
TEST(InterpreterPatternTest, EvaluateConstantsAndVariables) {
	Context ctx{{{"x", true}, {"y", false}}};

	Constant t(true), f(false);
	EXPECT_TRUE(t.interpret(ctx));
	EXPECT_FALSE(f.interpret(ctx));

	Variable vx("x"), vy("y");
	EXPECT_TRUE(vx.interpret(ctx));
	EXPECT_FALSE(vy.interpret(ctx));
}

/**
 * @brief Test composition of NOT, AND, and OR (AST-style).
 */
TEST(InterpreterPatternTest, ComposeNotAndAndOr) {
	Context ctx{{{"x", true}, {"y", false}, {"z", true}}};

	// Represents: !(x & y) | z
	std::unique_ptr<Expression> expr = std::make_unique<Or>(
		std::make_unique<Not>(
			std::make_unique<And>(
				std::make_unique<Variable>("x"),
				std::make_unique<Variable>("y")
			)
		),
		std::make_unique<Variable>("z")
	);

	EXPECT_TRUE(expr->interpret(ctx));
}

/**
 * @brief Unbound variables should throw std::out_of_range.
 */
TEST(InterpreterPatternTest, UnboundVariableThrows) {
	Context ctx{};
	Variable v("missing");
	EXPECT_THROW(v.interpret(ctx), std::out_of_range);
}

File: Behavioral/Interpreter/python/test_interpreter.py
import pytest
from interpreter import Expression, Constant, Variable, Not, And, Or, Parser

def test_constants_and_variables():
	"""
	Test evaluating constants and variable lookups.
	"""
	ctx = {"x": True, "y": False}
	assert Constant(True).interpret(ctx) is True
	assert Constant(False).interpret(ctx) is False
	assert Variable("x").interpret(ctx) is True
	assert Variable("y").interpret(ctx) is False

def test_composition_not_and_or():
	"""
	Test composition of NOT, AND, and OR.
	"""
	ctx = {"x": True, "y": False, "z": True}
	# !(x & y) | z
	expr = Or(
		Not(And(Variable("x"), Variable("y"))),
		Variable("z")
	)
	assert expr.interpret(ctx) is True

def test_operator_precedence():
	"""
	AND has higher precedence than OR; NOT has highest.
	"""
	ctx = {"x": False, "y": True, "z": True}
	# x | y & z  == x | (y & z) == True
	expr = Or(
		Variable("x"),
		And(Variable("y"), Variable("z"))
	)
	assert expr.interpret(ctx) is True

def test_unbound_variable_raises_keyerror():
	"""
	Variables absent from context should raise KeyError.
	"""
	ctx = {}
	with pytest.raises(KeyError):
		Variable("missing").interpret(ctx)

def test_parser_happy_path():
	"""
	Parse a valid expression and evaluate it.
	"""
	ctx = {"x": True, "y": False, "z": True}
	e = Parser("!(x & y) | z").parse()
	assert e.interpret(ctx) is True

def test_parser_errors():
	"""
	Invalid syntax should raise ValueError.
	"""
	with pytest.raises(ValueError):
		Parser("(").parse()
	with pytest.raises(ValueError):
		Parser("true garbage").parse()
	with pytest.raises(ValueError):
		Parser("! )").parse()
